Update select.js

redelschaap opened the pull request:
### Description
Fixed a bug where `options.reduce` was not available in case options was generated by an array with non-numeric or missing numeric keys in PHP. In that case a TypeError get thrown, because `options.reduce` is not a function of an object.

Uncaught TypeError: Unable to process binding "text: function (){return $col.getLabel($row()) }"
Message: options.reduce is not a function



### Manual testing scenarios
1. Have a select attribute where the options are being generated from an option array in PHP, where that array have non-numeric keys or missing numeric keys. For example:
```
$options = [
    1 => ['label' => 'Option A', 'value' => 1],
    3 => ['label' => 'Option C', 'value' => 3],
    4 => ['label' => 'Option D', 'value' => 4],
];
```
2. Load a grid where that attribute has an active column
3. The grid won't fully load (it loads until the first cell where that attribute is used) and JS throws an error:
```
Uncaught TypeError: Unable to process binding "text: function (){return $col.getLabel($row()) }"
Message: options.reduce is not a function
    at UiClass.flatOptions (select.js:58)
    at UiClass.getLabel (select.js:38)
    at UiClass.getLabel (wrapper.js:109)
    at text (eval at createBindingsStringEvaluator (knockout.js:2624), <anonymous>:3:69)
    at update (knockout.js:4260)
    at ko.dependentObservable.disposeWhenNodeIsRemoved (knockout.js:3004)
    at evaluateImmediate (knockout.js:1737)
    at Object.ko.computed.ko.dependentObservable (knockout.js:1946)
    at knockout.js:3002
    at Object.arrayForEach (knockout.js:151)
```

redelschaap committed:
Update select.js

Fixed a bug where options.reduce was not available in case options was generated by an array with non-numeric or missing numeric keys in PHP.

omiroshnichenko commented on app/code/Magento/Ui/view/base/web/js/grid/columns/select.js:
```diff
@@ -55,6 +55,10 @@ define([
         flatOptions: function (options) {
             var self = this;
 
+            if (!Array.isArray(options)) {
```
Please, use _.isArray function to check.

omiroshnichenko reviewed (commented):
Can you cover this with unit test? Test already created(dev/tests/js/jasmine/tests/app/code/Magento/Ui/base/js/grid/columns/select.test.js), but this case did not covered.

redelschaap committed:
Update select.js

Used `_.isArray` instead of `Array.isArray` as requested

redelschaap committed:
Update select.js

Used `_.isArray` instead of `Array.isArray` as requested

redelschaap committed:
Update select.test.js

Added a scenario where options are loaded as an object instead of an array. Options are generated in PHP. This is the case when the option array in PHP has non-numeric keys or has numeric keys but misses some numeric keys.

I noticed the opts test was deleted by someone in an earlier commit, but I think this is too important to remove from the test, so I added it again.

redelschaap reviewed (commented):
@omiroshnichenko I made the changes and covered it in the test file.

Wanted to add the note that the opt test is still present in the 2.1-develop branch, so I think it may be removed by mistake.

ishakhsuvarov commented:
@redelschaap Travis CI did not execute tests on this PR due to the error:
>  [`abuse detected: known offender (request looked fishy)`](https://travis-ci.org/magento/magento2/requests)

We will execute tests on this PR in the internal CICD, however you may want to contact Travis CI support to figure out what is wrong.
Thanks.

redelschaap commented:
Okay great. I guess they gave that error because I committed 4 or 5 times within half an hour. Is there anything I need to do now?

ishakhsuvarov commented:
@redelschaap It's ok on our side. We will let you know if there is anything specific found by internal CICD jobs.

ishakhsuvarov commented:
@redelschaap 
Looks like there are some errors regarding JS Unit tests implemented, here is partial output:
> Ui/js/grid/columns/select
> [exec]    getLabel method
> [exec]      - get label while options empty......✓
> [exec]      - get label for existed value......X
> [exec]        Expected '' to be 'b'. (1)
> [exec]      - get label for existed value in case the options are initialized as an object......X
> [exec]        Expected '' to be 'c'. (1)

I suggest you talking to Travis CI support to fix builds, so you could fix the errors.
Thanks

redelschaap committed:
Update select.test.js

Fix for jasmine tests

redelschaap commented:
@ishakhsuvarov I can't get grunt to work with jasmine tests, but when analyzing `select.js` and `column.js` this should do it. In the mean time I will try to reach Travis CI support.

redelschaap committed:
Update select.test.js

Changed file to re-run Travis test

redelschaap committed:
Update select.test.js

Fixed style issues

redelschaap commented:
@ishakhsuvarov Travis CI is working again on my account. Travis' tests for select.js succeeded, but two integration tests for PHP failed on some tests in Magento_Sales, but I did not change anything about that. In [Build #4295](https://travis-ci.org/magento/magento2/builds/237209451) you can see those integration tests succeeded, so I think Travis failed there.

ishakhsuvarov commented:
@redelschaap We are currently experiencing some issues with integration test failures. I've restarted build jobs, hope it's green this time.
Thank you for your effort.

redelschaap commented:
@ishakhsuvarov It did :)

magento-team commented:
@redelschaap thank you for your contribution

redelschaap commented:
You're welcome, happy to help!
